### PR TITLE
Initial revisions to support Collections as regular entities

### PR DESCRIPTION
**Pull request by kburke**
Initial revisions to support Collections as regular entities without disrupting existing Collections-only endpoints used for indexing.

### CODE REVIEW

File: src/app.py [hunk 1]
```diff
@@ -2982,6 +3072,18 @@ def get_prov_info_for_dataset(id):
             internal_dict[HEADER_PROCESSED_DATASET_UUID] = ",".join(processed_dataset_status_list)
             internal_dict[HEADER_PROCESSED_DATASET_UUID] = ",".join(processed_dataset_portal_url_list)
 
+    if include_samples:
+        dataset_samples = app_neo4j_queries.get_all_dataset_samples(neo4j_driver_instance, uuid)
+        logger.debug(f"dataset_samples={str(dataset_samples)}")
+        if 'all' in include_samples:
+            internal_dict[HEADER_DATASET_SAMPLES] = dataset_samples
```
Comment (maxsibilla) [inline src/app.py]: You have a check here for `include_samples` and handle whether its set to `all` or a specific UUID but you need to append the appropriate header to the parameter `headers`. If you run a GET request on localhost:5002/datasets/9d4363bbbf4fd4cc69d1827b4f929369/prov-info?include_samples=all you will get an error because when it creates a TSV, the number of columns for the headers is less than the columns for each row of data

File: src/schema/schema_constants.py [hunk 2]
```diff
@@ -14,3 +15,13 @@ class SchemaConstants(object):
...
+
+    DOI_BASE_URL = 'https://doi.org/'
+
+# Define an enumeration to classify an entity's visibility, which can be combined with
+# authorization info when verify operations on a request.
+class DataVisibilityEnum(Enum):
+    PUBLIC = SchemaConstants.ACCESS_LEVEL_PUBLIC
+    # Since initial release just requires public/non-public, add
+    # another entry indicating non-public.
+    NONPUBLIC = 'nonpublic'
```
Comment (maxsibilla) [inline src/schema/schema_constants.py:27]: What is the purpose of 'nonpublic' access level?

File: src/schema/schema_constants.py [hunk 3]
```diff
@@ -14,3 +15,13 @@ class SchemaConstants(object):
     ACCESS_LEVEL_PUBLIC = 'public'
     ACCESS_LEVEL_CONSORTIUM = 'consortium'
     ACCESS_LEVEL_PROTECTED = 'protected'
+
+    DOI_BASE_URL = 'https://doi.org/'
```
Comment (maxsibilla) [inline src/schema/schema_constants.py:19]: What is the purpose of this constant and shouldn't it be `https://dx.doi.org`?

File: src/app.py [hunk 4]
```diff
@@ -68,6 +71,38 @@
 # Suppress InsecureRequestWarning warning when requesting status on https with ssl cert verify disabled
 requests.packages.urllib3.disable_warnings(category = InsecureRequestWarning)
 
+####################################################################################################
```
Comment (maxsibilla) [inline src/app.py:74]: We have removed these because we utilize the library `atlas-consortia-commons` which contains these methods: https://github.com/x-atlas-consortia/commons/blob/main/atlas_consortia_commons/rest/__init__.py

File: src/app.py [hunk 5]
```diff
@@ -679,93 +807,31 @@ def get_entities_by_type(entity_type):
...
 
-    # The user_token is flask.Response on error
-    # Without token, the user can only access public collections, modify the collection result
-    # by only returning public datasets attached to this collection
-    if isinstance(user_token, Response):
-        # When the requested collection is not public, send back 401
-        if ('registered_doi' not in collection_dict) or ('doi_url' not in collection_dict):
-            # Require a valid token in this case
-            abort_unauthorized("The reqeusted collection is not public, a Globus token with the right access permission is required.")
+        complete_entities_list = schema_manager.get_complete_entities_list(token, entities_list, generated_properties_to_skip)
```
Comment (maxsibilla) [inline src/app.py]: There is no variable `generated_properties_to_skip`

File: src/app.py [hunk 6]
```diff
@@ -336,7 +374,121 @@ def get_ancestor_organs(id):
...
+    # since public collections don't require user token
+    token = get_internal_token()
+
+    # Get the entity dict from cache if exists
+    # Otherwise query against uuid-api and neo4j to get the entity dict if the id exists
+    collection_dict = query_target_entity(id, token)
+
+    # A bit validation
+    if collection_dict['entity_type'] != 'Collection':
+        bad_request_error("Target entity of the given id is not a collection")
```
Comment (maxsibilla) [inline src/app.py]: Replace `bad_request_error` with `abort_bad_req`

File: src/app.py [hunk 7]
```diff
@@ -336,7 +374,121 @@ def get_ancestor_organs(id):
...
+
+    # Response with the final result
+    return jsonify(normalized_complete_dict)
+
+def _get_entity_visibility(normalized_entity_type, entity_dict):
+    if normalized_entity_type not in schema_manager.get_all_entity_types():
+        logger.log( logging.ERROR
+                    ,f"normalized_entity_type={normalized_entity_type}"
+                     f" not recognized by schema_manager.get_all_entity_types().")
+        bad_request_error(f"'{normalized_entity_type}' is not a recognized entity type.")
```
Comment (maxsibilla) [inline src/app.py]: Repalce `bad_request_error` with `abort_bad_req`

File: src/app.py [hunk 8]
```diff
@@ -336,7 +374,121 @@ def get_ancestor_organs(id):
...
+    user_token = get_user_token(request)
+
+    # The user_token is flask.Response on error
+    # Without token, the user can only access public collections, modify the collection result
+    # by only returning public datasets attached to this collection
+    if isinstance(user_token, Response):
+        # When the requested collection is not public, send back 401
+        if ('registered_doi' not in collection_dict) or ('doi_url' not in collection_dict):
+            # Require a valid token in this case
+            http_unauthorized("The requested collection is not public, a Globus token with the right access permission is required.")
```
Comment (maxsibilla) [inline src/app.py]: Replace `http_unauthorized` with `abort_unauthorized`

File: src/app.py [hunk 9]
```diff
@@ -366,64 +518,44 @@ def get_entity_by_id(id):
...
-    # By now, either the entity is public accessible or the user token has the correct access level
-    # We'll need to return all the properties including those
-    # generated by `on_read_trigger` to have a complete result
-    # E.g., the 'next_revision_uuid' and 'previous_revision_uuid' being used below
-    # On entity retrieval, the 'on_read_trigger' doesn't really need a token
-    complete_dict = schema_manager.get_complete_entity_result(token, entity_dict)
+        # It's highly possible that there's no token provided
+        user_token = get_user_token(request)
 
-    # Additional handlings on dataset revisions
```
Comment (maxsibilla) [inline src/app.py:395]: Why was this removed? If you make a GET request to localhost:5002/entities/86784f0a82e12c95167c08124fbc949d without a bearer token the `next_revision_uuid` is present when it shouldn't be. Compare to https://entity-api.dev.sennetconsortium.org/entities/86784f0a82e12c95167c08124fbc949d

File: src/schema/schema_constants.py (hunk 2 above)
Comment (kburke) [reply src/schema/schema_constants.py:27]: The public/non-public notion of these constants (which are not "access levels") is the thumbs up or thumbs down on whether requested data can be shown to non-authenticated users.  This is a little different from the data_access_level, which is one component of deciding if something is public/non-public, determined by app.py's `_get_entity_visibility()`.

For example, the criteria for a Collection to be seen are numerous, including looping over each Dataset it contains:
[https://github.com/sennetconsortium/entity-api/blob/f98e2404e0bc2b10e5b23718b31c6fdb54017732/src/app.py#L465-L480](https://github.com/sennetconsortium/entity-api/blob/f98e2404e0bc2b10e5b23718b31c6fdb54017732/src/app.py#L465-L480)

File: src/app.py (hunk 1 above)
Comment (kburke) [reply src/app.py]: Not to disparage the fine people at GitHub, but can you confirm this is correctly attributed?  I don't see it at https://github.com/sennetconsortium/entity-api/compare/main...karlburke/collectionsAsEntities#diff-04791d82dd15fdd480f084d7ef65a10789fa5012cb7935f76080763444d48a00L2980-L3002
and it isn't anything I deliberately worked on.

I see it as a difference when I look at "Files Changed" on the PR, but Joe has told me this is some kind of "3 dot diff".  Perhaps it is useful for showing off at conferences I do not attend, but otherwise it seems like one more shortcoming of Git relative to direct source control, to a cranky old man like me!

Comment (maxsibilla) [reply src/app.py]: Gotcha, this must have sneaked past in a previous PR then. I will issue a fix and push that in with this PR

**Review by maxsibilla — CHANGES_REQUESTED** [review]

File: src/schema/schema_constants.py (hunk 3 above)
Comment (kburke) [reply src/schema/schema_constants.py:19]: When DOI is set, there are two fields on the JSON which must be provided together, 'doi_url' and 'registered_doi'.  Bill thinks it should be cut back to one in the future, but that is other work.  In the meantime, the two fields should have the same value, except doi_url should include the prefix specified by this constant.

The following query in HuBMAP Production Neo4j confirms the committed DOI_BASE_URL works for their existing data, and that nothing has a DOI in SenNet yet:
`MATCH (d:Dataset) WHERE d.doi_url IS NOT NULL RETURN DISTINCT LEFT(d.doi_url,15);`

Comment (yuanzhou) [reply src/schema/schema_constants.py:19]: @maxsibilla @kburke Early redirects were done through the domain [dx.doi.org](https://dx.doi.org/), but now almost all are done through [doi.org](https://doi.org/). The old one still works though, we should use the new one and no need to support the old URL. The base URL may change (less likely) but the actual DOI stays.

File: src/app.py (hunk 5 above)
Comment (kburke) [reply src/app.py]: Five minutes to confirm it was probably an accidental substitution in a section of code not changed by or tested for this feature. One minute to fix, ninety minutes to get into GitHub.

File: src/app.py (hunk 9 above)
Comment (kburke) [reply src/app.py:395]: The entire notion of retrieving a Dataset entity without authorization was changed by Bill.  The previous code would strip away certain elements of the entity if you were not authorized, but present you with the pieces that remained.  Now, you simply get the intact entity if authorized, or a 4xx error if not.

BTW, the old feature of a stripped-down Dataset cascaded into retrieval of Collections as well.  That is why it overlapped this work, and the old feature is removed in this branch.  (However, I can see some comments which are artifacts of the old feature, which I will remove.)

File: src/app.py (hunk 6 above)
Comment (kburke) [reply src/app.py]: Committed change which stripped local error handling routines, imported atlas_consortia_commons.rest routines, and replaced this call as directed.

File: src/app.py (hunk 7 above)
Comment (kburke) [reply src/app.py]: Committed change which stripped local error handling routines, imported atlas_consortia_commons.rest routines, and replaced this call as directed.

File: src/app.py (hunk 8 above)
Comment (kburke) [reply src/app.py]: Committed change which stripped local error handling routines, imported atlas_consortia_commons.rest routines, and replaced this call as directed.

File: src/app.py (hunk 4 above)
Comment (kburke) [reply src/app.py:74]: Committed change which stripped local error handling routines and imported `atlas_consortia_commons.rest `routines, Replaced calls as directed in other comments of this Discussion.

Also modified `schema/schema_manager.py` to import `atlas_consortia_commons.rest` routines, and to use `abort_bad_req()`, though I do not have a test to verify calls to `validate_target_entity_type_for_derivation()` and `validate_source_entity_type_for_derivation()` immediately.

File: src/app.py (hunk 7 above)
Comment (maxsibilla) [reply src/app.py]: @kburke I do not see these changes, https://github.com/sennetconsortium/entity-api/blob/98ab22b5a01df6a9b8626176bf2a6363130cb622/src/app.py#L74-L106

File: src/schema/provenance_schema.yaml [hunk 10]
```diff
@@ -238,15 +250,67 @@ ENTITIES:
...
       title:
         type: string
         description: "The title of the Collection."
         required_on_create: true
       description:
         type: string
         description: "Free text description of the collection"
+        required_on_create: true
       ###### Transient properties ######
+      dataset_uuids:
```
Comment (maxsibilla) [inline src/schema/provenance_schema.yaml:300]: Why do we need `dataset_uuids` and `datasets`?

File: src/app.py [hunk 11]
```diff
@@ -336,7 +341,121 @@ def get_ancestor_organs(id):
...
+        # Only return the public datasets attached to this Collection
+        if not user_in_globus_read_group(request):
+            complete_dict = get_complete_public_collection_dict(collection_dict)
+        else:
+            # We'll need to return all the properties including those
+            # generated by `on_read_trigger` to have a complete result
+            complete_dict = schema_manager.get_complete_entity_result(user_token, collection_dict)
+
+    # Will also filter the result based on schema
+    normalized_complete_dict = schema_manager.normalize_object_result_for_response( provenance_type='ENTITIES'
```
Comment (maxsibilla) [inline src/app.py:413]: Need to add `properties_to_include=['protocol_url']` to this method signature otherwise that property won't be grabbed from the Activity and returned with the rest of the entity properties. This is why a lot of the tests are failing:
`schema_manager.normalize_object_result_for_response('ENTITIES', complete_dict,
                                                                       properties_to_include=['protocol_url'])`

File: src/app.py [hunk 12]
```diff
@@ -366,64 +485,43 @@ def get_entity_by_id(id):
...
-                                                                       properties_to_include=['protocol_url'])
+    # We'll need to return all the properties including those generated by
+    # `on_read_trigger` to have a complete result e.g., the 'next_revision_uuid' and
+    # 'previous_revision_uuid' being used below.
+    if not user_authorized:
+        abort_forbidden(f"The requested {normalized_entity_type} has non-public data."
+                        f"  A Globus token with access permission is required.")
 
+    # Also normalize the result based on schema
+    final_result = schema_manager.normalize_object_result_for_response( provenance_type='ENTITIES'
```
Comment (maxsibilla) [inline src/app.py:523]: @kburke Need to add properties_to_include=['protocol_url'] to this method signature otherwise that property won't be grabbed from the Activity and returned with the rest of the entity properties. This is why a lot of the tests are failing:
schema_manager.normalize_object_result_for_response('ENTITIES', complete_dict, properties_to_include=['protocol_url'])

File: src/schema/provenance_schema.yaml (hunk 10 above)
Comment (kburke) [reply src/schema/provenance_schema.yaml:300]: `dataset_uuids` is for "writing" a Collection, more or less.  You can only give the UUID of an existing Dataset in this list, and the Dataset may be added to the new Collection.  You cannot specify JSON for a Dataset, for example, and expected it to be created and a part of your new Collection.

`datasets` is for "reading" a Collection.  The Datasets a Collection contains are part of the "transient" properties which are only read when they must be e.g. for a GET.

That is also reflected in provenance_schema.yaml `dataset_uuids` having `before*create` and `after*create` entries, and `datasets` having an `on_read*` entry.

**Review by maxsibilla — APPROVED** [review]
